src/coprocessor/dag/executor/scanner.rs

```diff
@@ -120,47 +111,79 @@ impl<S: Snapshot> Scanner<S> {
             );
         }
 
-        self.seek_key = match (self.scan_mode, self.scan_on) {
-            (ScanMode::Forward, _) => {
-                let mut seek_key = key.clone();
-                util::convert_to_prefix_next(&mut seek_key);
-                seek_key
-            }
-            (ScanMode::Backward, ScanOn::Table) => box_try!(truncate_as_row_key(&key)).to_vec(),
-            (ScanMode::Backward, ScanOn::Index) => key.clone(),
-            _ => unreachable!(),
-        };
+        // `Vec::clear()` produce 2 more instructions than `set_len(0)`, so we directly use
+        // `set_len()` here.
+        unsafe {
+            self.last_scanned_key.set_len(0);
+        }
+        self.last_scanned_key.extend_from_slice(key.as_slice());
 
         Ok(Some((key, value)))
     }
 
     pub fn start_scan(&self, range: &mut KeyRange) {
         assert!(!self.no_more);
-        let cur_seek_key = self.seek_key.clone();
-        match self.scan_mode {
-            ScanMode::Forward => range.set_start(cur_seek_key),
-            ScanMode::Backward => range.set_end(cur_seek_key),
-            _ => unreachable!(),
-        };
+        if self.last_scanned_key.is_empty() {
+            match self.scan_mode {
+                ScanMode::Forward => range.set_start(self.range.get_start().to_owned()),
+                ScanMode::Backward => range.set_end(self.range.get_end().to_owned()),
+                _ => unreachable!(),
+            }
+        } else {
+            match self.scan_mode {
+                ScanMode::Forward => {
+                    // In `stop_scan`, we will `next(last_scanned_key)`, so we don't need to next()
+                    // again here.
+                    range.set_start(self.last_scanned_key.clone())
+                }
+                ScanMode::Backward => {
+                    // In `stop_scan`, we have already truncated to row key, so we don't need to
+                    // do it again here.
+                    range.set_end(self.last_scanned_key.clone())
+                }
+                _ => unreachable!(),
+            }
+        }
     }
 
-    pub fn stop_scan(&self, range: &mut KeyRange) -> bool {
+    pub fn stop_scan(&mut self, range: &mut KeyRange) -> bool {
         if self.no_more {
             return false;
         }
-        let cur_seek_key = self.seek_key.clone();
-        match self.scan_mode {
-            ScanMode::Forward => range.set_end(cur_seek_key),
-            ScanMode::Backward => range.set_start(cur_seek_key),
-            _ => unreachable!(),
-        };
+        if self.last_scanned_key.is_empty() {
```

Comment:
When you call `stop_scan` and the previous `if self.no_more` don't return, I think the `last_scanned_key` must be not empty.

Reply:
Maybe there could be a call like start_scan(), end_scan()? (i.e. no scan during that at all)